File: stage1/solutions/python/challenge.py
import re
def encrypt(key, message):
    alphabet = [chr(i) for i in range(33, 126)]
    
    row_counter = 0
    output = ''
    shift = sum(ord(c) for c in key) # Shift each character of the sum of the ASCCI values of the key's characters
    for chunk in [message[i:i+len(key)] for i in range(0, len(message), len(key))]: # Split message in chunks of length key.size
        row_counter = row_counter + 1
        chunk = chunk[::-1] # Reverse chunk (note this is useless since we operate on chars :P)
        output = output + ''.join(alphabet[(alphabet.index(char) + shift) % len(alphabet)] for char in chunk)[::-1] + ' ' # Reverse chunk back (note this is useless since we operate on chars :P)
        if(row_counter % len(key) == 0): # Add a new line each key.size chunks
            output = output + '\n'

    return output.strip()

def decrypt(key, message):
    alphabet = [chr(i) for i in range(33, 126)]
    shift = sum(ord(c) for c in key)

    output = ''
    for chunk in re.compile(r"\s+").split(message):
        chunk = chunk[::-1] # Reverse chunk (note this is useless since we operate on chars :P)
        output_chunk = ''
        for char in chunk:
            output_char = (alphabet.index(char) - shift) % len(alphabet)
            if output_char < 0:
                output_char = len(alphabet) + output_char
            output_chunk = output_chunk + alphabet[output_char]
        output = output + output_chunk[::-1] # Reverse chunk back (note this is useless since we operate on chars :P)
    return output.strip()

if __name__ == '__main__':
    message = 'http://<fc:c2:de:34:42:7a>:3030/winner'
    key = 'wozniak'

    print('Encrypted secret is')
    encrypted = encrypt(key, message)
    
    print(encrypted);
    print('Decrypted secret is');
    print(decrypt(key, encrypted));

File: stage1/solutions/java/Challenge.java
public class Challenge
{
    public static char[] alphabet;
    public static String encrypt(String key, String message)
    {
        String [] chunks = message.split("(?<=\\G.{4})"); // Split message in chunks of length key.size
        
        int row_counter = 0;
        String output = "";
        int shift = 0;
        for(int i = 0; i < key.length(); i++) // Shift each character of the sum of the ASCCI values of the key's characters
            shift += key.charAt(i);

        for(String chunk : chunks)
        {
            chunk = new StringBuffer(chunk).reverse().toString(); // Reverse each chunk. Note that this is useless since we operate on chars
            row_counter++;
            String output_chunk = "";

            for(int i = 0; i < chunk.length(); i++)
            {
                int char_code = 0;
                for(int j = 0; j < 93; j++)
                {
                    if(Challenge.alphabet[j] == chunk.charAt(i))
                    {
                        char_code = j;
                        break;
                    }
                }

                int alphabet_index = (char_code + shift) % 93;
                output_chunk += alphabet[alphabet_index];
            }

            output += new StringBuffer(output_chunk).reverse().toString() + " "; // Reverse the chunk back
            if(row_counter % key.length() == 0) // Add a new line each key.size chunks
                output += '\n';
        }
        return output;
    }

    public static  String decrypt(String key, String message)
    {
        String [] chunks = message.split("\\s+"); // Split for whitespaces and new lines

        String output = "";
        int shift = 0;
        for(int i = 0; i < key.length(); i++) // Shift each character of the sum of the ASCCI values of the key's characters
            shift += key.charAt(i);

        for(String chunk : chunks)
        {
            chunk = new StringBuffer(chunk).reverse().toString();
            String output_chunk = "";
            for(int i = 0; i < chunk.length(); i++)
            {
                int char_code = 0;
                for(int j = 0; j < 93; j++)
                {
                    if(Challenge.alphabet[j] == chunk.charAt(i))
                    {
                        char_code = j;
                        break;
                    }
                }

                int output_char = (char_code - shift) % 93;
                if(output_char < 0)
                    output_char = 93 + output_char;
                char decrypted_char = alphabet[output_char];
                output_chunk += decrypted_char;
            }
            output += chunk = new StringBuffer(output_chunk).reverse().toString();;
        }

        return output;
    }

	public static void main(String [] args)
	{
        Challenge.alphabet = new char[93];
        for(int i = 0; i < 93; i++)
            alphabet[i] = (char) (i + 33);
    
		String message = "http://<fc:c2:de:34:42:7a>:3030/winner";
        String key = "wozniak";

        String encrypted = Challenge.encrypt(key, message);
        System.out.println("Encrypted secret is");
        System.out.println(encrypted);
        System.out.println("Decrypted secret is");
        System.out.println(Challenge.decrypt(key, encrypted));
	}
}
